baseplate/diagnostics/tracing.py

```diff
@@ -0,0 +1,185 @@
+            'ipv4': socket.gethostbyname(socket.gethostname()),
+        }
+        return {
+            'endpoint': endpoint_info,
+            'timestamp': timestamp,
+            'value': annotation_type
+        }
+
+    def _to_span_obj(self, annotations, binary_annotations):
+        span = {
+            "traceId": self.span.trace_id,
+            "name": self.span.name,
+            "id": self.span.id,
+            "timestamp": self.start,
+            "duration": self.elapsed,
+            "annotations": annotations,
+            "binary_annotations": binary_annotations
+        }
+
+        span['parentId'] = self.span.parent_id or 0
+        return span
+
+    def _serialize(self):
+        """Serialize span information into Zipkin-accepted format."""
+        annotations = []
+
+        annotations.append(
+            self._create_annotation(
+                ANNOTATIONS['CLIENT_RECEIVE'],
+                self.start
+            )
+        )
+
+        annotations.append(
+            self._create_annotation(
+                ANNOTATIONS['CLIENT_SEND'],
+                self.end
+            )
+        )
+
+        return self._to_span_obj(annotations, self.binary_annotations)
+
+    def record(self):
+        serialized = self._serialize()
+        self.recorder.send(serialized)
+
+
+class TraceServerSpanObserver(TraceSpanObserver):
+    def __init__(self, service_name, span, recorder):
+        self.service_name = service_name
+        self.span = span
+        self.recorder = recorder
+        super(TraceServerSpanObserver, self).__init__(service_name,
+                                                      span,
+                                                      recorder)
+
+    def on_start(self):
+        self.start = current_epoch_microseconds()
+
+    def on_child_span_created(self, child_span):
+        child_span_observer = TraceSpanObserver(self.recorder)
+        child_span.register(TraceSpanObserver(self.recorder))
+
+    def on_finish(self, exc_info):
+        self.end = current_epoch_microseconds()
+        self.elapsed = self.end - self.start
+        self.record()
+
+    def _serialize(self):
+        """Serialize span information into Zipkin-accepted format."""
+        annotations = []
+
+        annotations.append(self._create_annotation(ANNOTATIONS['SERVER_RECEIVE'],
+                                                   self.start))
+        annotations.append(self._create_annotation(ANNOTATIONS['SERVER_SEND'],
+                                                   self.end))
+
+        return self._to_span_obj(annotations, [])
+
+
+class NullRecorder(object):
+    def __init__(self):
+        pass
+
+    def send(self, span):
+        """Write a set of spans to debug log.
+        """
+        logger.debug(span)
+
+
+class RemoteRecorder(object):
+    def __init__(self, endpoint):
+        self.session = requests.Session()
+        self.endpoint = "http://%s/api/v1/spans" % endpoint
+
+    def send(self, span):
```

Comment:
ah, yeah, because the statsd thing happens inband. it might be worthwhile to change that, but it's less of a problem because it's udp and so there won't be much wait; mostly fire-and-forget except when dns goes wrong.